app/controllers/listings_controller.rb

```diff
@@ -8,6 +8,15 @@ def index
     @selected_search = params[:q]
     @selected_input = params.dig(:q, :selected_input) || "title"
     @listings = params[:q].present? ? @q.result(distinct: true) : []
+
+    if current_user && params[:q].present?
+      raw_q = begin
+        params[:q].to_unsafe_h
+      rescue
+        params[:q]
+      end
```

Comment:
Definitely should do [strong params](https://api.rubyonrails.org/classes/ActionController/StrongParameters.html) here. Strong params will convert this to a hash for you but will only allow users to provide details that you explicitly permit. With this, how it is currently, someone nefarious could submit some sketchy stuff in `params[:q]` and then when you go to check users subscriptions you could unknowingly run some evil code.